File: backend/src/main/java/fu/osms/customer/service/CustomerService.java
package fu.osms.customer.service;

import fu.osms.common.dto.PageResponse;
import fu.osms.customer.dto.request.CustomerRequest;
import fu.osms.customer.dto.response.CustomerResponse;
import fu.osms.customer.dto.response.CustomerStatsResponse;
import fu.osms.customer.dto.response.PageWithOrderCustomersResponse;
import fu.osms.customer.entity.Customer;

import java.util.UUID;

public interface CustomerService {

    CustomerResponse create(CustomerRequest request);

    CustomerResponse getById(UUID id);

    PageResponse<CustomerResponse> getAll(int page, int size, String search, String status, String gender);

    CustomerStatsResponse getStats();

    PageWithOrderCustomersResponse getPageWithOrderCustomers(int page, int size, String search, String status, String gender);

    CustomerResponse update(UUID id, CustomerRequest request);

    void delete(UUID id);

    /**
     * Sync KH ảo từ đơn hàng: quét tất cả order có customer_id IS NULL và gắn KH
     * vào từ buyer_name + buyer_phone. Dùng 1 lần khi boot để backfill.
     */
    int syncCustomersFromOrders();

    /**
     * Tìm KH theo fullName + phone, nếu không có thì tạo mới.
     * Dùng khi INSERT order mới có customer_id NULL.
     */
    Customer findOrCreateFromBuyer(String buyerName, String buyerPhone);
}

File: frontend/src/features/customer/services/customerService.js
import customerApi from '../../../api/customerApi';

const customerService = {
  getAll: async (page = 0, size = 20, search = '', status = 'ALL', gender = 'ALL') => {
    const data = await customerApi.getAll(page, size, search, status, gender);
    return data;
  },

  getById: async (id) => {
    const data = await customerApi.getById(id);
    return data;
  },

  create: async (data) => {
    const customer = await customerApi.create(data);
    return customer;
  },

  update: async (id, data) => {
    const customer = await customerApi.update(id, data);
    return customer;
  },

  delete: async (id) => {
    await customerApi.delete(id);
  },

  getStats: async () => {
    return await customerApi.getStats();
  },

  getPageWithOrderCustomers: async (page = 0, size = 20, search = '', status = 'ALL', gender = 'ALL') => {
    return await customerApi.getPageWithOrderCustomers(page, size, search, status, gender);
  },

  syncFromOrders: async () => {
    return await customerApi.syncFromOrders();
  },
};

export default customerService;